Create initial version of models

albexl opened the pull request.

albexl committed:
feat: Add basic version of models

albexl committed:
feat: Add some more methods

albexl committed:
test: Add more tests

albexl committed:
Add logic for the `_update_body` method

albexl committed:
Add scaffold of the Game class

josejorgers commented on src/model/snake.py:
```diff
@@ -0,0 +1,61 @@
+
+    def _create_body(self, position: Tuple[int, int]) -> List[Tuple[int, int]]:
+        body = [position]
+        body.extend((body[i - 1][0] - 10, position[1]) for i in range(1, 4))
```
What does 10 means? Shouldn't it be a parameter or a constant?

albexl replied:
I just took it from the GeeksForGeeks example. I think it should be `1` instead of `10`. Maybe they use `10` in that tutorial for some specific reason due to coupling. I will investigate.

josejorgers commented on src/model/snake.py:
```diff
@@ -0,0 +1,61 @@
+
+    def move(self):
+        """Updates the position of the head."""
+        if self.direction == Direction.RIGHT:
+            self.position = (self.position[0] + 10, self.position[1])
+        elif self.direction == Direction.LEFT:
+            self.position = (self.position[0] - 10, self.position[1])
+        elif self.direction == Direction.UP:
+            self.position = (self.position[0], self.position[1] - 10)
+        else:
+            self.position = (self.position[0], self.position[1] + 10)
```
Same here with 10

albexl replied:
Same answer

josejorgers commented on src/model/fruit.py:
```diff
@@ -0,0 +1,11 @@
+class Fruit:
+    """Representation of the Fruit entity.
+
+    Probably best to use `dataclass` here.
```
agree

josejorgers commented on src/game_state/game.py:
```diff
@@ -0,0 +1,19 @@
+class Game:
```
In this folder, my idea was to have classes like `UserAction` and `GameState`. I think this class `Game` could be like a state machine.

- `UserAction` is any action that the user can perform
- `GameState` is a `Board` along with other functionalities like `is_final` function
- `Game` can be initialized with a `Board`, it contains a property `current_state` and a method `go_to` that receives a `UserAction` and modifies the `current_state`

I think this is a good abstraction in the form of a state machine. We don't have to talk about state machines explicitly but I think it is good for development purposes.

albexl replied:
This approach is great. Agree.

josejorgers commented on src/model/board.py:
```diff
@@ -0,0 +1,6 @@
+class Board:
```
Here we can have a board composed of a matrix of tiles where each tile has type `typing.Union[Fruit, Snake, None]`

Maybe it is worth it to declare an alias like `Tile = Union[Fruit, Snake, None]`

albexl replied:
I haven't used aliases very much before. I will look into them also.

josejorgers reviewed (commented):
I don't know too much about `pygame`, though maybe what I'm saying is not useful but here we go:

1. I think the main logic should be approached from a state machine POV. The way I picture it is included in the code comments
2. Then maybe it would be good to add a subfolder `rendering` with all the code needed to render a `GameState`

This is how I see the big picture. There are a few details that I commented on in the code, but everything looks good.

I'm keeping the PR unmerged for now so we can have time to agree on an approach to follow but from the code perspective it is good for me.

albexl committed:
First functional version

Missing:
- Still some functionality needs to be placed in the proper place.
- State machine mechanism.

josejorgers commented on src/game_state/game.py:
```diff
@@ -1,19 +1,112 @@
+    def touch_body(self) -> bool:
+        return any(self.snake.position == block for block in self.snake.body[1:])
+
+    def game_over(self):
```
Although it is very uncommon, it is possible to win the game. This happens when all the board is occupied by the snake

josejorgers commented on src/game_state/game.py:
```diff
@@ -1,19 +1,112 @@
+            else:
+                self.snake.trim()
+
+            if not self.fruit_spawn:
```
Maybe it would be useful to have a constructor that only receives the size of the board and generates the random position.

josejorgers replied:
Or maybe a factory IDK

josejorgers commented on src/model/fruit.py:
```diff
@@ -1,11 +1,9 @@
 
 
+@dataclass
 class Fruit:
```
Mmmmm I see, this is a dataclass. Then maybe it would be good to have a helper method that generates an instance of this class in a random position from the window dimensions

josejorgers commented on src/model/snake.py:
```diff
@@ -14,14 +14,18 @@ class Direction(Enum):
+        self.body = self._create_body(position, size)
 
-    def _create_body(self, position: Tuple[int, int]) -> List[Tuple[int, int]]:
+    def _create_body(
```
Maybe adding a comment explaining what the initial position of the snake is can help to understand the code better. I think the snake begins in a horizontal position starting from `position` and going down in the `x`-axis, but it took me some time to understand it

josejorgers reviewed (commented):
It looks good. I added some comments, not merging yet since you told me you'll add some changes

albexl committed:
Add final adjustments for the first version

josejorgers commented on src/game_state/game.py:
```diff
@@ -9,16 +17,14 @@
     def __init__(self):
         pygame.init()
         pygame.display.set_caption("Snake")
+        self.snake_speed = 15
+        self.window_x = 720
+        self.window_y = 480
```
These should be at least optional parameters. In the future, we could make the window size configurable or let the user select the difficulty level.

albexl replied:
Ack

josejorgers commented on src/game_state/game.py:
```diff
@@ -30,23 +36,50 @@ def __init__(self):
 
-    def touch_boundaries(self) -> bool:
+    def _touch_boundaries(self) -> bool:
+        """Checks whether the snake's head is touching the
+        boundaries of the board.
+
+        Returns:
+            bool: True in case the snake's head is touching the boundaries, False otherwise.
+        """
```
This is something I have thought of but never mentioned. IDK if the comments should be in Spanish, not too sure but maybe yes. What you think?

albexl replied:
They should be in Spanish, yes.

albexl replied:
But I don't know if the code details should be in Spanish also

josejorgers commented on src/game_state/game.py:
```diff
@@ -30,23 +36,50 @@ def __init__(self):
+            * Snake's head touches body.
+
+        Returns:
+            bool: True if any condition is met, False otherwise.
```
We should add the condition that the snake occupies the whole board. This rarely happens but if we later create an automatic snake that plays perfectly it should achieve it.

albexl replied:
Ack

josejorgers reviewed (approved):
This is good for the first version. Let's review my proposals later and see if we incorporate the changes I commented.
